{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5a3c8a47-8201-4fe6-ba87-7f9d4dc35441",
   "metadata": {},
   "source": [
    "# LOADING DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eef3cf75-2579-4b0b-a83d-51d5060f1dd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                                                                                                                          | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>in_tissue</th>\n",
       "      <th>array_row</th>\n",
       "      <th>array_col</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>destripe_factor</th>\n",
       "      <th>n_counts_adjusted</th>\n",
       "      <th>labels_he</th>\n",
       "      <th>labels_he_expanded</th>\n",
       "      <th>labels_gex</th>\n",
       "      <th>labels_joint</th>\n",
       "      <th>labels_joint_source</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>s_002um_02587_02503-1</th>\n",
       "      <td>1</td>\n",
       "      <td>2587</td>\n",
       "      <td>2503</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.369561</td>\n",
       "      <td>12.195529</td>\n",
       "      <td>290012</td>\n",
       "      <td>290012</td>\n",
       "      <td>42453</td>\n",
       "      <td>290012</td>\n",
       "      <td>primary</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_01674_00710-1</th>\n",
       "      <td>1</td>\n",
       "      <td>1674</td>\n",
       "      <td>710</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.368149</td>\n",
       "      <td>12.148933</td>\n",
       "      <td>0</td>\n",
       "      <td>221472</td>\n",
       "      <td>35534</td>\n",
       "      <td>221472</td>\n",
       "      <td>primary</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_02498_02808-1</th>\n",
       "      <td>1</td>\n",
       "      <td>2498</td>\n",
       "      <td>2808</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.321744</td>\n",
       "      <td>10.617562</td>\n",
       "      <td>269059</td>\n",
       "      <td>269059</td>\n",
       "      <td>0</td>\n",
       "      <td>269059</td>\n",
       "      <td>primary</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_00952_00136-1</th>\n",
       "      <td>1</td>\n",
       "      <td>952</td>\n",
       "      <td>136</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.706108</td>\n",
       "      <td>23.301573</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17982</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_00564_00910-1</th>\n",
       "      <td>1</td>\n",
       "      <td>564</td>\n",
       "      <td>910</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.091521</td>\n",
       "      <td>3.020180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       in_tissue  array_row  array_col  n_counts  \\\n",
       "s_002um_02587_02503-1          1       2587       2503      16.0   \n",
       "s_002um_01674_00710-1          1       1674        710       5.0   \n",
       "s_002um_02498_02808-1          1       2498       2808       6.0   \n",
       "s_002um_00952_00136-1          1        952        136       9.0   \n",
       "s_002um_00564_00910-1          1        564        910       1.0   \n",
       "\n",
       "                       destripe_factor  n_counts_adjusted  labels_he  \\\n",
       "s_002um_02587_02503-1         0.369561          12.195529     290012   \n",
       "s_002um_01674_00710-1         0.368149          12.148933          0   \n",
       "s_002um_02498_02808-1         0.321744          10.617562     269059   \n",
       "s_002um_00952_00136-1         0.706108          23.301573          0   \n",
       "s_002um_00564_00910-1         0.091521           3.020180          0   \n",
       "\n",
       "                       labels_he_expanded  labels_gex  labels_joint  \\\n",
       "s_002um_02587_02503-1              290012       42453        290012   \n",
       "s_002um_01674_00710-1              221472       35534        221472   \n",
       "s_002um_02498_02808-1              269059           0        269059   \n",
       "s_002um_00952_00136-1                   0       17982             0   \n",
       "s_002um_00564_00910-1                   0           0             0   \n",
       "\n",
       "                      labels_joint_source   sample_id  \n",
       "s_002um_02587_02503-1             primary  IMMUNEX001  \n",
       "s_002um_01674_00710-1             primary  IMMUNEX001  \n",
       "s_002um_02498_02808-1             primary  IMMUNEX001  \n",
       "s_002um_00952_00136-1                none  IMMUNEX001  \n",
       "s_002um_00564_00910-1                none  IMMUNEX001  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:19<00:00, 19.41s/it]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 18315.74it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>in_tissue</th>\n",
       "      <th>array_row</th>\n",
       "      <th>array_col</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>destripe_factor</th>\n",
       "      <th>n_counts_adjusted</th>\n",
       "      <th>labels_he</th>\n",
       "      <th>labels_he_expanded</th>\n",
       "      <th>labels_gex</th>\n",
       "      <th>labels_joint</th>\n",
       "      <th>labels_joint_source</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>s_002um_00303_02786-1</th>\n",
       "      <td>1</td>\n",
       "      <td>303</td>\n",
       "      <td>2786</td>\n",
       "      <td>21.576923</td>\n",
       "      <td>0.653846</td>\n",
       "      <td>21.576923</td>\n",
       "      <td>0</td>\n",
       "      <td>17873</td>\n",
       "      <td>6014</td>\n",
       "      <td>17873</td>\n",
       "      <td>primary</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_01259_00369-1</th>\n",
       "      <td>1</td>\n",
       "      <td>1259</td>\n",
       "      <td>369</td>\n",
       "      <td>25.610031</td>\n",
       "      <td>0.776062</td>\n",
       "      <td>25.610031</td>\n",
       "      <td>0</td>\n",
       "      <td>220570</td>\n",
       "      <td>38743</td>\n",
       "      <td>220570</td>\n",
       "      <td>primary</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_02580_02931-1</th>\n",
       "      <td>1</td>\n",
       "      <td>2580</td>\n",
       "      <td>2931</td>\n",
       "      <td>5.322421</td>\n",
       "      <td>0.161285</td>\n",
       "      <td>5.322421</td>\n",
       "      <td>0</td>\n",
       "      <td>310649</td>\n",
       "      <td>0</td>\n",
       "      <td>310649</td>\n",
       "      <td>primary</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_00598_02061-1</th>\n",
       "      <td>1</td>\n",
       "      <td>598</td>\n",
       "      <td>2061</td>\n",
       "      <td>11.974747</td>\n",
       "      <td>0.362871</td>\n",
       "      <td>11.974747</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_00531_03162-1</th>\n",
       "      <td>1</td>\n",
       "      <td>531</td>\n",
       "      <td>3162</td>\n",
       "      <td>8.550374</td>\n",
       "      <td>0.259102</td>\n",
       "      <td>8.550374</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       in_tissue  array_row  array_col   n_counts  \\\n",
       "s_002um_00303_02786-1          1        303       2786  21.576923   \n",
       "s_002um_01259_00369-1          1       1259        369  25.610031   \n",
       "s_002um_02580_02931-1          1       2580       2931   5.322421   \n",
       "s_002um_00598_02061-1          1        598       2061  11.974747   \n",
       "s_002um_00531_03162-1          1        531       3162   8.550374   \n",
       "\n",
       "                       destripe_factor  n_counts_adjusted  labels_he  \\\n",
       "s_002um_00303_02786-1         0.653846          21.576923          0   \n",
       "s_002um_01259_00369-1         0.776062          25.610031          0   \n",
       "s_002um_02580_02931-1         0.161285           5.322421          0   \n",
       "s_002um_00598_02061-1         0.362871          11.974747          0   \n",
       "s_002um_00531_03162-1         0.259102           8.550374          0   \n",
       "\n",
       "                       labels_he_expanded  labels_gex  labels_joint  \\\n",
       "s_002um_00303_02786-1               17873        6014         17873   \n",
       "s_002um_01259_00369-1              220570       38743        220570   \n",
       "s_002um_02580_02931-1              310649           0        310649   \n",
       "s_002um_00598_02061-1                   0           0             0   \n",
       "s_002um_00531_03162-1                   0           0             0   \n",
       "\n",
       "                      labels_joint_source   sample_id  n_genes  \n",
       "s_002um_00303_02786-1             primary  IMMUNEX001        9  \n",
       "s_002um_01259_00369-1             primary  IMMUNEX001       26  \n",
       "s_002um_02580_02931-1             primary  IMMUNEX001        4  \n",
       "s_002um_00598_02061-1                none  IMMUNEX001        6  \n",
       "s_002um_00531_03162-1                none  IMMUNEX001        4  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import squidpy as sq\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "from anndata import AnnData\n",
    "import scipy.sparse as sp\n",
    "import anndata as ad\n",
    "import os\n",
    "# os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'\n",
    "\n",
    "sample_ids = [f\"{i:03d}\" for i in range(1, 19)]\n",
    "sample_ids = ['001']\n",
    "\n",
    "# List to store individual adatas\n",
    "adatas = []\n",
    "adata_dict = {}\n",
    "spatial_dict = {}\n",
    "\n",
    "\n",
    "# Loop through each sample\n",
    "for sid in tqdm(sample_ids):\n",
    "    sample_id = f\"IMMUNEX{sid}\"\n",
    "    path = f\"/scratch/Projects/IMMUNEX/segmentation/bin2cell/bin2cell_output/{sample_id}/adata_processed.h5ad\"\n",
    "    \n",
    "    if os.path.exists(path):\n",
    "        adata_sample = sc.read(path)\n",
    "        adata_sample.var_names_make_unique()\n",
    "        adata_sample.raw = adata_sample\n",
    "        adata_sample.uns['sample_id'] = sample_id\n",
    "        adata_sample.obs['sample_id'] = sample_id  # Add sample ID to obs for tracking\n",
    "        display(adata_sample.obs.head())\n",
    "\n",
    "        # Count total UMIs per bin\n",
    "        adata_sample.obs['n_counts'] = np.array(adata_sample.X.sum(axis=1)).flatten()\n",
    "        # Count number of genes per bin\n",
    "        adata_sample.obs['n_genes'] = np.array((adata_sample.X > 0).sum(axis=1)).flatten()\n",
    "\n",
    "        \n",
    "        # Save .uns['spatial'] if it exists\n",
    "        if 'spatial' in adata_sample.uns:\n",
    "            spatial_dict[sample_id] = adata_sample.uns['spatial']\n",
    "            \n",
    "        adata_dict[sample_id] = adata_sample\n",
    "\n",
    "        # adatas.append(adata_sample)\n",
    "    else:\n",
    "        print(f\"File not found for sample {sample_id}: {path}\")\n",
    "\n",
    "    \n",
    "# Concatenate all samples into one AnnData object\n",
    "# adata = ad.concat(adatas, label='batch', keys=sample_ids, index_unique='-')\n",
    "adata = ad.concat(adata_dict, label='sample_id', merge='unique')\n",
    "adata.uns['spatial'] = spatial_dict\n",
    "\n",
    "for sid in tqdm(adata.uns['spatial']):\n",
    "    nested_key = list(adata.uns['spatial'][sid].keys())[0]  # e.g., 'Visium_NSCLC_IMMUNEX018'\n",
    "    adata.uns['spatial'][sid] = adata.uns['spatial'][sid][nested_key]\n",
    "\n",
    "adata.obs.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6b89d7fa-7b05-4f44-ac46-c84f75cd4b3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>in_tissue</th>\n",
       "      <th>array_row</th>\n",
       "      <th>array_col</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>destripe_factor</th>\n",
       "      <th>n_counts_adjusted</th>\n",
       "      <th>labels_he</th>\n",
       "      <th>labels_he_expanded</th>\n",
       "      <th>labels_gex</th>\n",
       "      <th>labels_joint</th>\n",
       "      <th>...</th>\n",
       "      <th>Average_fragment_size</th>\n",
       "      <th>Concentratio_pg_per_µL</th>\n",
       "      <th>Concentration finale  (ng/µL)</th>\n",
       "      <th>Final_concentration_nM</th>\n",
       "      <th>Dual _Index</th>\n",
       "      <th>Volume_available_µL</th>\n",
       "      <th>Volume_send_µL</th>\n",
       "      <th>Sample_code</th>\n",
       "      <th>Position_on_the_plaque</th>\n",
       "      <th>Sea_Dual_Index_TS_Set_A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2587</td>\n",
       "      <td>2503</td>\n",
       "      <td>12.195529</td>\n",
       "      <td>0.369561</td>\n",
       "      <td>12.195529</td>\n",
       "      <td>290012</td>\n",
       "      <td>290012</td>\n",
       "      <td>42453</td>\n",
       "      <td>290012</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>46.6</td>\n",
       "      <td>2.33</td>\n",
       "      <td>14.0</td>\n",
       "      <td>G3</td>\n",
       "      <td>around_24</td>\n",
       "      <td>22</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>A1</td>\n",
       "      <td>SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1674</td>\n",
       "      <td>710</td>\n",
       "      <td>12.148933</td>\n",
       "      <td>0.368149</td>\n",
       "      <td>12.148933</td>\n",
       "      <td>0</td>\n",
       "      <td>221472</td>\n",
       "      <td>35534</td>\n",
       "      <td>221472</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>46.6</td>\n",
       "      <td>2.33</td>\n",
       "      <td>14.0</td>\n",
       "      <td>G3</td>\n",
       "      <td>around_24</td>\n",
       "      <td>22</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>A1</td>\n",
       "      <td>SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2498</td>\n",
       "      <td>2808</td>\n",
       "      <td>10.617562</td>\n",
       "      <td>0.321744</td>\n",
       "      <td>10.617562</td>\n",
       "      <td>269059</td>\n",
       "      <td>269059</td>\n",
       "      <td>0</td>\n",
       "      <td>269059</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>46.6</td>\n",
       "      <td>2.33</td>\n",
       "      <td>14.0</td>\n",
       "      <td>G3</td>\n",
       "      <td>around_24</td>\n",
       "      <td>22</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>A1</td>\n",
       "      <td>SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>952</td>\n",
       "      <td>136</td>\n",
       "      <td>23.301573</td>\n",
       "      <td>0.706108</td>\n",
       "      <td>23.301573</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17982</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>46.6</td>\n",
       "      <td>2.33</td>\n",
       "      <td>14.0</td>\n",
       "      <td>G3</td>\n",
       "      <td>around_24</td>\n",
       "      <td>22</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>A1</td>\n",
       "      <td>SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>564</td>\n",
       "      <td>910</td>\n",
       "      <td>3.020180</td>\n",
       "      <td>0.091521</td>\n",
       "      <td>3.020180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>252</td>\n",
       "      <td>46.6</td>\n",
       "      <td>2.33</td>\n",
       "      <td>14.0</td>\n",
       "      <td>G3</td>\n",
       "      <td>around_24</td>\n",
       "      <td>22</td>\n",
       "      <td>IMMUNEX001</td>\n",
       "      <td>A1</td>\n",
       "      <td>SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   in_tissue  array_row  array_col   n_counts  destripe_factor  \\\n",
       "0          1       2587       2503  12.195529         0.369561   \n",
       "1          1       1674        710  12.148933         0.368149   \n",
       "2          1       2498       2808  10.617562         0.321744   \n",
       "3          1        952        136  23.301573         0.706108   \n",
       "4          1        564        910   3.020180         0.091521   \n",
       "\n",
       "   n_counts_adjusted  labels_he  labels_he_expanded  labels_gex  labels_joint  \\\n",
       "0          12.195529     290012              290012       42453        290012   \n",
       "1          12.148933          0              221472       35534        221472   \n",
       "2          10.617562     269059              269059           0        269059   \n",
       "3          23.301573          0                   0       17982             0   \n",
       "4           3.020180          0                   0           0             0   \n",
       "\n",
       "   ... Average_fragment_size Concentratio_pg_per_µL  \\\n",
       "0  ...                   252                   46.6   \n",
       "1  ...                   252                   46.6   \n",
       "2  ...                   252                   46.6   \n",
       "3  ...                   252                   46.6   \n",
       "4  ...                   252                   46.6   \n",
       "\n",
       "   Concentration finale  (ng/µL) Final_concentration_nM  Dual _Index  \\\n",
       "0                           2.33                   14.0           G3   \n",
       "1                           2.33                   14.0           G3   \n",
       "2                           2.33                   14.0           G3   \n",
       "3                           2.33                   14.0           G3   \n",
       "4                           2.33                   14.0           G3   \n",
       "\n",
       "  Volume_available_µL Volume_send_µL Sample_code Position_on_the_plaque  \\\n",
       "0           around_24             22  IMMUNEX001                     A1   \n",
       "1           around_24             22  IMMUNEX001                     A1   \n",
       "2           around_24             22  IMMUNEX001                     A1   \n",
       "3           around_24             22  IMMUNEX001                     A1   \n",
       "4           around_24             22  IMMUNEX001                     A1   \n",
       "\n",
       "                     Sea_Dual_Index_TS_Set_A  \n",
       "0  SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT  \n",
       "1  SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT  \n",
       "2  SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT  \n",
       "3  SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT  \n",
       "4  SI-TS-G3,CCAGACACGG,AGAAAGCGGT,ACCGCTTTCT  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD11</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL17</th>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHN1</th>\n",
       "      <td>ENSG00000187583</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>984</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ids    feature_types  genome  n_cells\n",
       "OR4F5    ENSG00000186092  Gene Expression  GRCh38       18\n",
       "SAMD11   ENSG00000187634  Gene Expression  GRCh38      328\n",
       "NOC2L    ENSG00000188976  Gene Expression  GRCh38     1941\n",
       "KLHL17   ENSG00000187961  Gene Expression  GRCh38     1126\n",
       "PLEKHN1  ENSG00000187583  Gene Expression  GRCh38      984"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load clinical data\n",
    "clinical_df = pd.read_csv(\"/home/mounim/rawdata/IMMUNEX/data/VisiumHD_18_2024_NSCLC.csv\")\n",
    "\n",
    "# Preview\n",
    "clinical_df['sample_id'] = clinical_df['HE_image_name'].str[:10]\n",
    "\n",
    "adata.obs = adata.obs.merge(clinical_df, on='sample_id', how='left')\n",
    "\n",
    "display(adata.obs.head())\n",
    "display(adata.var.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "15c5c265-cd4b-4bf2-b599-8dcffbc13e52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9613112, 18322)\n",
      "(36966, 18322)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mounim/miniconda3/envs/enact/lib/python3.10/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "# get a sample for testing\n",
    "# 1. Filter to crop region (spatial coordinates)\n",
    "# Define crop bounds\n",
    "x_min, x_max = 100, 300\n",
    "y_min, y_max = 100, 300\n",
    "\n",
    "# Assuming 'array_row' and 'array_col' are in `adata.obs`\n",
    "crop_mask = (\n",
    "    (adata.obs['array_row'] >= x_min) & (adata.obs['array_row'] <= x_max) &\n",
    "    (adata.obs['array_col'] >= y_min) & (adata.obs['array_col'] <= y_max)\n",
    ")\n",
    "\n",
    "adata_crop = adata[crop_mask].copy()\n",
    "\n",
    "print(adata.shape)\n",
    "print(adata_crop.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eda81a11-5e69-4a28-8caa-a99fc343f3b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mounim/miniconda3/envs/enact/lib/python3.10/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1889 × 18322\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'n_counts', 'destripe_factor', 'n_counts_adjusted', 'labels_he', 'labels_he_expanded', 'labels_gex', 'labels_joint', 'labels_joint_source', 'sample_id', 'n_genes', 'Manip_Visium', 'DV200', 'HE_image_name', 'Cytasimage_path', 'TLS_status', 'area', 'slide', 'Average_fragment_size', 'Concentratio_pg_per_µL', 'Concentration finale  (ng/µL)', 'Final_concentration_nM', 'Dual _Index', 'Volume_available_µL', 'Volume_send_µL', 'Sample_code', 'Position_on_the_plaque', 'Sea_Dual_Index_TS_Set_A'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'\n",
       "    uns: 'spatial'\n",
       "    obsm: 'spatial', 'spatial_cropped_150_buffer'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_min, x_max = 1000, 1050\n",
    "y_min, y_max = 1000, 1050\n",
    "\n",
    "# Assuming 'array_row' and 'array_col' are in `adata.obs`\n",
    "crop_mask = (\n",
    "    (adata.obs['array_row'] >= x_min) & (adata.obs['array_row'] <= x_max) &\n",
    "    (adata.obs['array_col'] >= y_min) & (adata.obs['array_col'] <= y_max)\n",
    ")\n",
    "\n",
    "adata_crop = adata[crop_mask].copy()\n",
    "adata_crop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0947de25-eb66-4c09-bac3-a4537b9875db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>array_row</th>\n",
       "      <th>array_col</th>\n",
       "      <th>isSegmented</th>\n",
       "      <th>isNuclei</th>\n",
       "      <th>isGExCell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2587</td>\n",
       "      <td>2503</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1674</td>\n",
       "      <td>710</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2498</td>\n",
       "      <td>2808</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>952</td>\n",
       "      <td>136</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>564</td>\n",
       "      <td>910</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   array_row  array_col  isSegmented  isNuclei  isGExCell\n",
       "0       2587       2503         True      True      False\n",
       "1       1674        710         True     False      False\n",
       "2       2498       2808         True      True      False\n",
       "3        952        136         True     False       True\n",
       "4        564        910        False     False      False"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get bins and segmentation status \n",
    "adata.obs[['labels_he', 'labels_he_expanded', 'labels_gex']] = adata.obs[['labels_he', 'labels_he_expanded', 'labels_gex']].fillna(0)\n",
    "\n",
    "# Add isSegmented column\n",
    "adata.obs['isSegmented'] = (\n",
    "    (adata.obs['labels_he'] != 0) |\n",
    "    (adata.obs['labels_he_expanded'] != 0) |\n",
    "    (adata.obs['labels_gex'] != 0)\n",
    ")\n",
    "\n",
    "# Add isNuclei column\n",
    "adata.obs['isNuclei'] = adata.obs['labels_he'] != 0\n",
    "\n",
    "# Add isGExCell column\n",
    "adata.obs['isGExCell'] = (adata.obs['labels_gex'] != 0) & (adata.obs['labels_he_expanded'] == 0)\n",
    "\n",
    "bins_annotation = adata.obs[['array_row','array_col','isSegmented', 'isNuclei', 'isGExCell']]\n",
    "bins_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171ec63f-f305-41aa-95be-4efd8abbfe05",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3d186811-6fe1-46c9-9271-d62efb9333f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mounim/miniconda3/envs/enact/lib/python3.10/site-packages/numba/np/ufunc/parallel.py:371: NumbaWarning: The TBB threading layer requires TBB version 2021 update 6 or later i.e., TBB_INTERFACE_VERSION >= 12060. Found TBB_INTERFACE_VERSION = 12050. The TBB threading layer is disabled.\n",
      "  warnings.warn(problem)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "target = 'labels_he'  # or 'isSegmented' or 'isGExCell'\n",
    "\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=1000)\n",
    "adata_hvg = adata[:, adata.var['highly_variable']]\n",
    "X = adata_hvg.X  # sparse matrix (safe to keep sparse)\n",
    "y = adata.obs[target].astype(int)\n",
    "\n",
    "# Keep only classes with ≥2 samples\n",
    "value_counts = y.value_counts()\n",
    "valid_classes = value_counts[value_counts >= 2].index\n",
    "mask = y.isin(valid_classes)\n",
    "\n",
    "# Convert mask to indices for sparse matrix slicing\n",
    "valid_indices = np.where(mask)[0]\n",
    "\n",
    "# Subset both X and y safely\n",
    "X = X[valid_indices]\n",
    "y = y.iloc[valid_indices]\n",
    "\n",
    "\n",
    "# Now safe to stratify\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, y, test_size=0.2, random_state=42, stratify=y\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bda7002-17a9-4ca0-ae05-626bb1af13ac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff088b4e-015c-4f0f-94ae-799954f6a2dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lightgbm import LGBMClassifier\n",
    "\n",
    "clf = LGBMClassifier(n_estimators=100, random_state=42)\n",
    "clf.fit(X_train, y_train)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9de3e3f-d4f5-4e7e-84f2-48ec6d090f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # RandomForest can accept sparse matrix directly in recent versions\n",
    "# clf = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=222, verbose =1)\n",
    "# clf.fit(X_train, y_train)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a40c276a-7c28-4d70-89cd-2a42897c867b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lightgbm import LGBMClassifier\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "# Predict on test set\n",
    "y_pred = clf.predict(X_test)\n",
    "\n",
    "# Compute accuracy\n",
    "acc = accuracy_score(y_test, y_pred)\n",
    "print(f\"Test Accuracy: {acc:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a25d51e1-ba14-4a3f-80a3-cfbd07d27d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import accuracy_score, roc_curve, auc\n",
    "\n",
    "# Predict probabilities\n",
    "y_pred_proba = clf.predict_proba(X_test)[:, 1]  # Probability for class 1\n",
    "y_pred = clf.predict(X_test)\n",
    "\n",
    "# Accuracy\n",
    "acc = accuracy_score(y_test, y_pred)\n",
    "print(f\"Test Accuracy: {acc:.4f}\")\n",
    "\n",
    "# ROC curve\n",
    "fpr, tpr, thresholds = roc_curve(y_test, y_pred_proba)\n",
    "roc_auc = auc(fpr, tpr)\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(6, 6))\n",
    "plt.plot(fpr, tpr, label=f\"ROC curve (AUC = {roc_auc:.2f})\", linewidth=2)\n",
    "plt.plot([0, 1], [0, 1], linestyle='--', color='gray')\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.title(\"ROC Curve\")\n",
    "plt.legend(loc=\"lower right\")\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9ade490-c269-4832-95a4-228584c8c521",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
