{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "282cf8b3-dc0f-4596-ba9d-4eee0f7f0c27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bin2cell_output  bin2cell_output_he0.01_gex0.05  bin2cell_output_he0005_gex005\n"
     ]
    }
   ],
   "source": [
    "! ls /scratch/Projects/IMMUNEX/segmentation/bin2cell/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ffcc00cf-a035-41ce-8930-a976ef6e34d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/Banksy_py/banksy/cluster_methods.py:18: UserWarning: \n",
      "No rpy2 installed. BANKSY will run, but mclust will not work.\n",
      "Note: you can still use the default leiden option for clustering.\n",
      "Install rpy2 and R in your conda environment if you want to use mclust for clustering.\n",
      "\n",
      "  warnings.warn(warn_str)\n",
      "2025-07-10 22:48:31.879986: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2025-07-10 22:48:31.889659: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2025-07-10 22:48:31.901504: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2025-07-10 22:48:31.905043: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2025-07-10 22:48:31.914715: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-07-10 22:48:32.649692: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(\"/opt/Banksy_py\")\n",
    "import banksy as banksy\n",
    "import time; import random\n",
    "\n",
    "from banksy.initialize_banksy import initialize_banksy\n",
    "from banksy.run_banksy import run_banksy_multiparam\n",
    "from banksy_utils.color_lists import spagcn_color\n",
    "from banksy_utils.plot_utils import plot_qc_hist, plot_cell_positions\n",
    "from banksy_utils.load_data import load_adata, display_adata\n",
    "from banksy_utils.filter_utils import filter_cells\n",
    "from banksy_utils.filter_utils import normalize_total, filter_hvg, print_max_min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b2c55b7-2499-4144-8557-3420e1406822",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 10287064 × 18270\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'n_counts', 'destripe_factor', 'n_counts_adjusted', 'labels_he', 'labels_he_expanded', 'labels_gex', 'labels_joint', 'labels_joint_source'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'\n",
       "    uns: 'bin2cell', 'spatial', 'sample_id'\n",
       "    obsm: 'spatial', 'spatial_cropped_150_buffer'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "from anndata import AnnData\n",
    "import scipy.sparse as sp\n",
    "\n",
    "# Load the data\n",
    "sample_id = \"IMMUNEX002\"\n",
    "\n",
    "adata = sc.read(f\"/scratch/Projects/IMMUNEX/segmentation/bin2cell/bin2cell_output_he0.01_gex0.05/{sample_id}/adata_processed.h5ad\")\n",
    "adata.raw = adata\n",
    "adata.uns['sample_id'] = sample_id\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96828f40-c839-4962-9d24-cb47c6c13954",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>in_tissue</th>\n",
       "      <th>array_row</th>\n",
       "      <th>array_col</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>destripe_factor</th>\n",
       "      <th>n_counts_adjusted</th>\n",
       "      <th>labels_he</th>\n",
       "      <th>labels_he_expanded</th>\n",
       "      <th>labels_gex</th>\n",
       "      <th>labels_joint</th>\n",
       "      <th>labels_joint_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>s_002um_02587_02503-1</th>\n",
       "      <td>1</td>\n",
       "      <td>2587</td>\n",
       "      <td>2503</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>21.411765</td>\n",
       "      <td>0</td>\n",
       "      <td>123152</td>\n",
       "      <td>22606</td>\n",
       "      <td>123152</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_01674_00710-1</th>\n",
       "      <td>1</td>\n",
       "      <td>1674</td>\n",
       "      <td>710</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.233977</td>\n",
       "      <td>6.083413</td>\n",
       "      <td>0</td>\n",
       "      <td>168030</td>\n",
       "      <td>34839</td>\n",
       "      <td>168030</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_02498_02808-1</th>\n",
       "      <td>1</td>\n",
       "      <td>2498</td>\n",
       "      <td>2808</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.315715</td>\n",
       "      <td>8.208591</td>\n",
       "      <td>119256</td>\n",
       "      <td>119256</td>\n",
       "      <td>0</td>\n",
       "      <td>119256</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_00952_00136-1</th>\n",
       "      <td>1</td>\n",
       "      <td>952</td>\n",
       "      <td>136</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.143678</td>\n",
       "      <td>3.735632</td>\n",
       "      <td>330594</td>\n",
       "      <td>330594</td>\n",
       "      <td>0</td>\n",
       "      <td>330594</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_002um_00564_00910-1</th>\n",
       "      <td>1</td>\n",
       "      <td>564</td>\n",
       "      <td>910</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.528161</td>\n",
       "      <td>13.732175</td>\n",
       "      <td>0</td>\n",
       "      <td>291153</td>\n",
       "      <td>0</td>\n",
       "      <td>291153</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       in_tissue  array_row  array_col  n_counts  \\\n",
       "s_002um_02587_02503-1          1       2587       2503      28.0   \n",
       "s_002um_01674_00710-1          1       1674        710       2.0   \n",
       "s_002um_02498_02808-1          1       2498       2808       4.0   \n",
       "s_002um_00952_00136-1          1        952        136       1.0   \n",
       "s_002um_00564_00910-1          1        564        910       5.0   \n",
       "\n",
       "                       destripe_factor  n_counts_adjusted  labels_he  \\\n",
       "s_002um_02587_02503-1         0.823529          21.411765          0   \n",
       "s_002um_01674_00710-1         0.233977           6.083413          0   \n",
       "s_002um_02498_02808-1         0.315715           8.208591     119256   \n",
       "s_002um_00952_00136-1         0.143678           3.735632     330594   \n",
       "s_002um_00564_00910-1         0.528161          13.732175          0   \n",
       "\n",
       "                       labels_he_expanded  labels_gex  labels_joint  \\\n",
       "s_002um_02587_02503-1              123152       22606        123152   \n",
       "s_002um_01674_00710-1              168030       34839        168030   \n",
       "s_002um_02498_02808-1              119256           0        119256   \n",
       "s_002um_00952_00136-1              330594           0        330594   \n",
       "s_002um_00564_00910-1              291153           0        291153   \n",
       "\n",
       "                      labels_joint_source  \n",
       "s_002um_02587_02503-1             primary  \n",
       "s_002um_01674_00710-1             primary  \n",
       "s_002um_02498_02808-1             primary  \n",
       "s_002um_00952_00136-1             primary  \n",
       "s_002um_00564_00910-1             primary  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the metadata\n",
    "metadata = pd.read_csv(\"~/rawdata/IMMUNEX/data/VisiumHD_18_2024_NSCLC.csv\")\n",
    "\n",
    "metadata_sample = metadata[metadata['Sample_code'] == sample_id]\n",
    "adata.uns[\"sample_metadata\"] = metadata_sample.iloc[0].to_dict()\n",
    "\n",
    "display(adata.obs.head()) # Display the first few rows of the observation data (cell metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "132c827f-21dc-49e3-b2d4-90610a7cb4b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d22509a0-1c39-4eaf-8146-69f3408ae1c5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4a0cb19e-b9d0-497b-b4d0-e56e07be7221",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 1543059 × 18270\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'n_counts', 'destripe_factor', 'n_counts_adjusted', 'labels_he', 'labels_he_expanded', 'labels_gex', 'labels_joint', 'labels_joint_source'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'\n",
      "    uns: 'bin2cell', 'spatial', 'sample_id', 'sample_metadata'\n",
      "    obsm: 'spatial', 'spatial_cropped_150_buffer'\n"
     ]
    }
   ],
   "source": [
    "# Sampling adata \n",
    "# Set seed for reproducibility\n",
    "np.random.seed(42)\n",
    "\n",
    "# Compute number of cells to sample\n",
    "n_obs = adata.n_obs\n",
    "n_sample = int(n_obs * 0.01)\n",
    "n_sample = int(n_obs * 0.15)\n",
    "\n",
    "# Randomly choose indices\n",
    "sample_indices = np.random.choice(n_obs, size=n_sample, replace=False)\n",
    "\n",
    "# Subset the AnnData object\n",
    "adata_sampled = adata[sample_indices].copy()\n",
    "\n",
    "# Inspect new shape\n",
    "print(adata_sampled)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f23414-e747-45cd-9179-412aac4ebc5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sampled.obs['cell_id'] = adata_sampled.obs['labels_joint'].astype(str)\n",
    "adata_cells = adata_sampled[adata_sampled.obs['labels_joint'] > 0].copy()\n",
    "adata_cells.var['mt'] = adata_sampled.var_names.str.upper().str.startswith('MT-')\n",
    "\n",
    "# Ensure matrix is dense or convert to DataFrame\n",
    "X = adata_cells.X.toarray() if sp.issparse(adata_cells.X) else adata_cells.X\n",
    "df = pd.DataFrame(X, columns=adata_cells.var_names, index=adata_cells.obs['cell_id'])\n",
    "\n",
    "# Compute mean expression per cell_id\n",
    "grouped_expr = df.groupby(level=0).mean()\n",
    "\n",
    "# Create a new AnnData object\n",
    "adata_cells_mean = AnnData(X=sp.csr_matrix(grouped_expr.values))\n",
    "adata_cells_mean.obs_names = grouped_expr.index\n",
    "adata_cells_mean.var_names = grouped_expr.columns\n",
    "\n",
    "adata_cells_mean.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a75bb033-0c10-4acc-8538-7cefaaf60ccd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5946ca3-2012-4e3e-b1c5-11be54512f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and 'cell_id' is the grouping key used to generate grouped_expr\n",
    "# 1. Get a mapping from cell_id → average spatial position\n",
    "cell_coords = adata_sampled.obs[['cell_id']].copy()\n",
    "cell_coords['x'] = adata_sampled.obsm['spatial'][:, 0]\n",
    "cell_coords['y'] = adata_sampled.obsm['spatial'][:, 1]\n",
    "\n",
    "# 2. Average spatial position per cell\n",
    "mean_coords = cell_coords.groupby('cell_id')[['x', 'y']].mean()\n",
    "\n",
    "# 3. Match to your grouped_expr.index (cell IDs)\n",
    "mean_coords = mean_coords.loc[grouped_expr.index]\n",
    "\n",
    "# 4. Assign to obsm\n",
    "adata_cells_mean.obsm['spatial'] = mean_coords.values\n",
    "\n",
    "# Copy the spatial dictionary to cdata\n",
    "library_id = list(adata_sampled.uns['spatial'].keys())[0]\n",
    "adata_cells_mean.uns['spatial'] = {\n",
    "    library_id: adata_sampled.uns['spatial'][library_id]\n",
    "}\n",
    "adata_cells = adata_cells_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e116033-91dd-4a88-825c-71ee99a26cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add metadata\n",
    "# Step 1: Copy gene metadata (var) for the grouped expression matrix\n",
    "adata_cells.var = adata_sampled.var.loc[grouped_expr.columns].copy()\n",
    "adata_cells.var_names_make_unique()\n",
    "\n",
    "# Step 2: Aggregate metadata per reconstructed cell\n",
    "meta = adata_sampled.obs.groupby('cell_id').agg({\n",
    "    'n_counts': 'mean',\n",
    "})\n",
    "# - Number of bins per cell (count of occurrences)\n",
    "meta['n_bins'] = adata_sampled.obs['cell_id'].value_counts()\n",
    "meta.index.name = 'cell_id'\n",
    "\n",
    "# Step 3: Reorder metadata to match cell ordering in grouped_expr\n",
    "meta = meta.loc[grouped_expr.index]\n",
    "\n",
    "# Step 4: Assign aggregated metadata as the new obs for reconstructed cells\n",
    "adata_cells.obs = meta\n",
    "\n",
    "# Step 5: Copy over select entries from uns\n",
    "for k in ['sample_id', 'sample_metadata', 'bin2cell']:\n",
    "    if k in adata.uns:\n",
    "        adata_cells.uns[k] = adata.uns[k]\n",
    "\n",
    "# Step 6: Copy full spatial metadata\n",
    "adata_cells.uns['spatial'] = adata.uns['spatial']\n",
    "\n",
    "# Done\n",
    "adata_cells\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84ad26b4-8546-42b8-9b52-b08b6e6e639f",
   "metadata": {},
   "source": [
    "## Cropped viz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f26d8e-e9a3-4857-b220-7684c04aa284",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "crop_axis_fraction = 0.05  # how much to crop around the center\n",
    "\n",
    "# Get spatial coordinates\n",
    "spatial_coords = adata_cells.obsm['spatial'].copy()\n",
    "\n",
    "# Compute coordinate bounds and center\n",
    "x_min, x_max = np.min(spatial_coords[:, 0]), np.max(spatial_coords[:, 0])\n",
    "y_min, y_max = np.min(spatial_coords[:, 1]), np.max(spatial_coords[:, 1])\n",
    "center_x = (x_min + x_max) / 2\n",
    "center_y = (y_min + y_max) / 2\n",
    "\n",
    "# Define crop bounds around center\n",
    "x0 = center_x - (x_max - x_min) * crop_axis_fraction / 2\n",
    "x1 = center_x + (x_max - x_min) * crop_axis_fraction / 2\n",
    "y0 = center_y - (y_max - y_min) * crop_axis_fraction / 2\n",
    "y1 = center_y + (y_max - y_min) * crop_axis_fraction / 2\n",
    "\n",
    "# Find cells within crop\n",
    "within_crop = (\n",
    "    (spatial_coords[:, 0] >= x0) & (spatial_coords[:, 0] <= x1) &\n",
    "    (spatial_coords[:, 1] >= y0) & (spatial_coords[:, 1] <= y1)\n",
    ")\n",
    "adata_cropped = adata_cells[within_crop, :].copy()\n",
    "\n",
    "# Normalize n_bins to define spot sizes (you can adjust scaling)\n",
    "sizes = adata_cropped.obs['n_bins']\n",
    "sizes_normalized = 10 * (sizes / sizes.max())  # scale to 0–20\n",
    "\n",
    "# Plot\n",
    "sc.pl.spatial(\n",
    "    adata_cropped,\n",
    "    color=['n_counts',None],\n",
    "    size=sizes_normalized,\n",
    "    cmap='Spectral_r',\n",
    "    alpha_img=0.6,\n",
    "    show=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78fe1127-4b81-4891-a508-f0638aaa0cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "crop_axis_fraction = 0.33  # how much to crop around the center\n",
    "\n",
    "# Get spatial coordinates\n",
    "spatial_coords = adata_cells.obsm['spatial'].copy()\n",
    "\n",
    "# Compute coordinate bounds and center\n",
    "x_min, x_max = np.min(spatial_coords[:, 0]), np.max(spatial_coords[:, 0])\n",
    "y_min, y_max = np.min(spatial_coords[:, 1]), np.max(spatial_coords[:, 1])\n",
    "center_x = (x_min + x_max) / 2\n",
    "center_y = (y_min + y_max) / 2\n",
    "\n",
    "# Define crop bounds around center\n",
    "x0 = center_x - (x_max - x_min) * crop_axis_fraction / 2\n",
    "x1 = center_x + (x_max - x_min) * crop_axis_fraction / 2\n",
    "y0 = center_y - (y_max - y_min) * crop_axis_fraction / 2\n",
    "y1 = center_y + (y_max - y_min) * crop_axis_fraction / 2\n",
    "\n",
    "# Find cells within crop\n",
    "within_crop = (\n",
    "    (spatial_coords[:, 0] >= x0) & (spatial_coords[:, 0] <= x1) &\n",
    "    (spatial_coords[:, 1] >= y0) & (spatial_coords[:, 1] <= y1)\n",
    ")\n",
    "adata_cropped = adata_cells[within_crop, :].copy()\n",
    "\n",
    "# Normalize n_bins to define spot sizes (you can adjust scaling)\n",
    "sizes = adata_cropped.obs['n_bins']\n",
    "sizes_normalized = 50 * (sizes / sizes.max())  # scale to 0–20\n",
    "\n",
    "# Plot\n",
    "sc.pl.spatial(\n",
    "    adata_cropped,\n",
    "    color=['n_counts',None],\n",
    "    size=sizes_normalized,\n",
    "    cmap='Spectral_r',\n",
    "    alpha_img=0.6,\n",
    "    show=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0afb726-09da-4055-8cd2-95c421730955",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata_cells, inplace=True)\n",
    "adata_cells.obs['n_bins'] = adata_cells.obs['n_bins'].astype(float)\n",
    "sns.scatterplot(x=adata_cells.obs['n_bins'], y=adata_cells.obs['total_counts'])\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f58a59c-e402-4a56-9365-4ab49ff216d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4701da88-f4c6-4d97-afb4-2ef6183e7a82",
   "metadata": {},
   "source": [
    "## QC Adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5507d7c-be7e-45a6-9b4a-f7a555687715",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"/opt/Banksy_py\")\n",
    "import banksy as banksy\n",
    "import time; import random\n",
    "\n",
    "from banksy.initialize_banksy import initialize_banksy\n",
    "from banksy.run_banksy import run_banksy_multiparam\n",
    "from banksy_utils.color_lists import spagcn_color\n",
    "from banksy_utils.plot_utils import plot_qc_hist, plot_cell_positions\n",
    "from banksy_utils.load_data import load_adata, display_adata\n",
    "from banksy_utils.filter_utils import filter_cells\n",
    "from banksy_utils.filter_utils import normalize_total, filter_hvg, print_max_min\n",
    "\n",
    "import scipy.sparse\n",
    "if not scipy.sparse.issparse(adata_cells.X):\n",
    "    print('converting to sparse')\n",
    "    adata_cells.X = scipy.sparse.csr_matrix(adata_cells.X)\n",
    "    # adata.raw = adata.X\n",
    "display_adata(adata_cells)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7ef06e5-e4d0-498b-af8d-3615da6f5d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# QC\n",
    "adata_cells.var_names_make_unique()\n",
    "adata_cells.var[\"mt\"] = adata_cells.var_names.str.startswith(\"MT-\")\n",
    "\n",
    "sc.pp.calculate_qc_metrics(adata_cells, \n",
    "                           qc_vars=[\"mt\"], \n",
    "                           log1p=True, \n",
    "                           inplace=True)\n",
    "\n",
    "# bin options for fomratting histograms\n",
    "# Here, we set 'auto' for 1st figure, 80 bins for 2nd figure. and so on\n",
    "hist_bin_options = ['auto', 50, 80, 100]\n",
    "\n",
    "plot_qc_hist(adata_cells, \n",
    "         total_counts_cutoff = 800, # for visualization\n",
    "         n_genes_high_cutoff = 1000, # for visualization\n",
    "         n_genes_low_cutoff = 0, # for visualization\n",
    "         bin_options = hist_bin_options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ae5e53-58d8-4e1e-a855-ad18fef0e2d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initial filter of cells by their cell counts, MT count and gene counts\n",
    "adata_filtred = adata_cells.copy()\n",
    "\n",
    "adata_filtred = filter_cells(adata_filtred, \n",
    "             min_count=3, \n",
    "             max_count=500, \n",
    "             MT_filter=20, \n",
    "             gene_filter=100)\n",
    "\n",
    "plot_qc_hist(adata_filtred,\n",
    "         total_counts_cutoff = 800, # for visualization\n",
    "         n_genes_high_cutoff = 1000, # for visualization\n",
    "         n_genes_low_cutoff = 0, # for visualization\n",
    "        bin_options = hist_bin_options)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf1acaa0-9324-4e67-b427-5ad8ad61ca60",
   "metadata": {},
   "source": [
    "## Normalize and scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65ba9a8c-da07-4fe6-9452-2da770c03e77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify highly variable genes — modify filter_hvg to skip log1p inside\n",
    "adata_normalized = adata_filtred.copy()\n",
    "adata_normalized = normalize_total(adata_normalized)\n",
    "sc.pp.log1p(adata_normalized)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38dd4791-a9a4-4e29-8d08-2f24d2a7a812",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pp.highly_variable_genes(adata_normalized, n_top_genes=5000, flavor='seurat')\n",
    "\n",
    "nsclc_marker_genes = sorted(list(set([\n",
    "    # T cells\n",
    "    'CD3D', 'CD3E', 'CD3G', 'CD2', 'CD7', 'TRAC', 'JUNB', 'S100A4', 'CD52', 'PFN1P1',\n",
    "    'CD81', 'EEF1B2P3', 'CXCR4', 'CREM', 'IL32', 'TGIF1',\n",
    "\n",
    "    # CD4+ T cells\n",
    "    'CD4', 'IL7R', 'CCR7', 'SELL', 'FOXP3', 'CTLA4', 'TIGIT',\n",
    "\n",
    "    # CD8+ T cells\n",
    "    'CD8A', 'CD8B', 'GZMA', 'GZMB', 'PRF1', 'LAG3', 'PDCD1',\n",
    "\n",
    "    # Regulatory T cells\n",
    "    'IL2RA', 'IKZF2',\n",
    "\n",
    "    # B cells\n",
    "    'CD19', 'MS4A1', 'CD79A', 'CD79B', 'IGHD', 'IGHM', 'CD22', 'CD24',\n",
    "    'CD74', 'HMGA1', 'CD52', 'PTPRC', 'HLA-DRA', 'CXCR4', 'SPCS3', 'LTB', 'IGKC',\n",
    "\n",
    "    # Plasma cells\n",
    "    'MZB1', 'JCHAIN', 'SDC1', 'XBP1',\n",
    "\n",
    "    # NK cells\n",
    "    'NCAM1', 'KLRB1', 'KLRD1', 'NKG7', 'GNLY', 'PRF1', 'FGFBP2',\n",
    "    'IL32', 'FHL2', 'IL2RG', 'CD69', 'HOPX',\n",
    "\n",
    "    # Dendritic cells\n",
    "    'ITGAX', 'CD1C', 'CLEC9A', 'CD83', 'LAMP3', 'CCR7', 'HLA-DRA', 'BATF3',\n",
    "\n",
    "    # Monocytes / Macrophages\n",
    "    'CD14', 'CD68', 'LYZ', 'FCN1', 'S100A8', 'S100A9', 'ITGAM', 'MRC1', 'CD163',\n",
    "    'NOS2', 'IL1B', 'TNF', 'ARG1', 'CD206', 'IL10',\n",
    "\n",
    "    # Neutrophils\n",
    "    'FCGR3B', 'CSF3R', 'ELANE', 'MPO', 'CEACAM8',\n",
    "\n",
    "    # Endothelial\n",
    "    'PECAM1', 'VWF', 'CDH5', 'KDR', 'CLDN5', 'ESAM', 'ENG', 'CD34', 'PROM1', 'PDPN',\n",
    "    'TEK', 'FLT1', 'VCAM1', 'PTPRC', 'MCAM', 'ICAM1', 'FLT4',\n",
    "\n",
    "    # Fibroblasts / CAFs\n",
    "    'PDGFRA', 'PDGFRB', 'COL1A1', 'COL1A2', 'COL3A1', 'COL5A2', 'ACTA2', 'TAGLN', 'FAP', 'POSTN',\n",
    "    'TCF21', 'FN',\n",
    "\n",
    "    # Pericytes / Smooth muscle\n",
    "    'RGS5', 'MYL9', 'MYLK', 'FHL2', 'ITGA1', 'EHD2', 'OGN', 'SNCG', 'FABP4',\n",
    "\n",
    "    # Epithelial / Alveolar / Cancer cells\n",
    "    'EPCAM', 'KRT8', 'KRT18', 'KRT19', 'SFTPC', 'SFTPB', 'AGER', 'PDPN', 'KRT7', 'KRT5', 'KRT14',\n",
    "    'TP63', 'NGFR', 'MUC1', 'MUC16', 'BIRC5', 'MYC', 'SOX2', 'CCND1', 'CDKN2A', 'PD-L1',\n",
    "    'EGFR', 'KRAS', 'TP53', 'TTF-1', 'NKX2-1', 'CEACAM5', 'CDH1', 'CLDN1',\n",
    "\n",
    "    # Club/secretory cells\n",
    "    'SCGB1A1', 'SCGB3A2', 'MUC5B',\n",
    "\n",
    "    # Ciliated cells\n",
    "    'FOXJ1', 'TPPP3', 'PIFO', 'DNAH5',\n",
    "\n",
    "    # Goblet cells\n",
    "    'MANF', 'KRT7', 'AQP3', 'AGR2', 'BACE2', 'TFF3', 'PHGR1', 'MUC4', 'MUC13', 'GUCA2A',\n",
    "\n",
    "    # Enterocytes\n",
    "    'CD55', 'ELF3', 'PLIN2', 'GSTM3', 'KLF5', 'CBR1', 'APOA1', 'CA1', 'PDHA1', 'EHF',\n",
    "\n",
    "    # Enteroendocrine\n",
    "    'NUCB2', 'FABP5', 'CPE', 'ALCAM', 'GCG', 'SST', 'CHGB', 'IAPP', 'CHGA', 'ENPP2',\n",
    "\n",
    "    # Crypt cells / stem-like\n",
    "    'HOPX', 'SLC12A2', 'MSI1', 'SMOC2', 'OLFM4', 'ASCL2', 'PROM1', 'BMI1', 'EPHB2', 'LRIG1',\n",
    "\n",
    "    # Cancer stem cells\n",
    "    'CD44', 'PROM1', 'ALDH1A1', 'ITGA6',\n",
    "\n",
    "    # Immune checkpoints / tumor markers\n",
    "    'PDCD1', 'CTLA4', 'HAVCR2', 'TIGIT', 'LAG3', 'CD274', 'PDCD1LG2',\n",
    "    'HLA-A', 'HLA-B', 'HLA-C', 'HLA-DQA1', 'HLA-DPB1'\n",
    "])))\n",
    "\n",
    "hvg_genes = adata_normalized.var_names[adata_normalized.var['highly_variable']].tolist()\n",
    "combined_genes = list(set(hvg_genes + nsclc_marker_genes))\n",
    "genes_to_keep = [g for g in combined_genes if g in adata_normalized.var_names]\n",
    "adata_normalized = adata_normalized[:, genes_to_keep]\n",
    "adata_normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ce7d31d-1e54-44e4-a21d-aa50dd4a5461",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52ad98e7-a229-4694-9387-323c4df7b90a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata_normalized, svd_solver='arpack',n_comps=100)\n",
    "# “elbow” — the point where additional PCs contribute little variance — and use that number in n_pcs.\n",
    "sc.pl.pca_variance_ratio(adata_normalized, log=True, n_pcs=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce9672fc-4bc7-4d81-a5e6-a0f1508c799f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Parameters\n",
    "NNei = 30\n",
    "res = 0.01\n",
    "nPCs = 30\n",
    "\n",
    "# 1. Compute neighborhood graph\n",
    "sc.pp.neighbors(adata_normalized, n_neighbors=NNei, n_pcs=nPCs)\n",
    "\n",
    "print(f'Clustering with resolution = {res}')\n",
    "# 2. Run Leiden clustering\n",
    "leiden_key = f'nn_{NNei} leiden_{res}'\n",
    "sc.tl.leiden(\n",
    "    adata_normalized,\n",
    "    resolution=res,\n",
    "    key_added=leiden_key,\n",
    ")\n",
    "print(f'Clustering done, ploting...')\n",
    "\n",
    "sc.pl.embedding(\n",
    "    adata_normalized,\n",
    "    basis=\"spatial\",\n",
    "    color=f'nn_{NNei} leiden_{res}',\n",
    "    size=5,\n",
    "    alpha=0.6,\n",
    "    title=f\"Spatial clusters (Leiden res={res})\",\n",
    "    show=False\n",
    ")\n",
    "plt.gca().invert_yaxis()\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c141c52-56a9-4b46-b116-24dac1d45ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'Generating UMAP for resolution = {res}')\n",
    "\n",
    "# 3. Compute UMAP embedding\n",
    "sc.tl.umap(adata_normalized)\n",
    "\n",
    "print(f'UMAP done, ploting...')\n",
    "\n",
    "# 5. Plot UMAP colored by clustering result\n",
    "sc.pl.umap(\n",
    "    adata_normalized,\n",
    "    color=leiden_key,\n",
    "    title=f'nn_{NNei} Leiden {res}',\n",
    "    size=10  # adjust based on dataset\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "315eb150-1ff2-4199-ac3d-ede8984da427",
   "metadata": {},
   "source": [
    "## Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acd64c70-b002-4f02-b1a1-64bfc0fbe90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_dict = {\n",
    "    'B cell': [\n",
    "        'CD19', 'CD20', 'CD79A', 'CD79B','IGHM', 'MS4A1'\n",
    "    ],\n",
    "    'Basal cell': [\n",
    "        'TP63', 'NGFR', 'MST1R'\n",
    "    ],\n",
    "    'CD4+ T cell': ['CD4'],\n",
    "    'CD8+ T cell': ['CD8A'],\n",
    "    'Cancer cell': [\n",
    "        'EGFR', 'EPCAM', 'TTF-1', 'KRAS', 'TP53', 'NKX2-1',\n",
    "        'TP63', 'CEACAM5', 'LOX'\n",
    "    ],\n",
    "    'Dendritic cell': [\n",
    "        'CCL17', 'CD141', 'CD1A', 'CD1C', 'CD303', 'CD83'\n",
    "    ],\n",
    "    'Endothelial cell': [\n",
    "        'CA4', 'CD31', 'CD34', 'CDH5', 'CLDN5', 'COL15A1', 'CPE',\n",
    "        'DKK2', 'EDNRB', 'PECAM1', 'VWF', 'STEAP1', 'KDR', 'VEGFA'\n",
    "    ],\n",
    "    'Epithelial cell': [\n",
    "        'AGER', 'CCDC153', 'CD326', 'CDH1', 'E-cadherin', 'EPCAM',\n",
    "        'FOXJ1', 'KRT18', 'KRT19', 'KRT5', 'KRT7', 'KRT8', 'MUC5B',\n",
    "        'SCGB3A2', 'SFTPB', 'SFTPC', 'TPPP3'\n",
    "    ],\n",
    "    'Fibroblast': ['COL1A1', 'COL1A2', 'PDGFRA', 'aSMA','FAP'],\n",
    "    'Macrophage': [\n",
    "        'APOE', 'CCL18', 'CCL2', 'CD11b', 'CD11c', 'CD14', 'CD163',\n",
    "        'CD64', 'CD68', 'CD80', 'CD86', 'CSF1R', 'FN1', 'ITGAM',\n",
    "        'MRC1', 'MS4A4A', 'SPP1', 'VCAN'\n",
    "    ],\n",
    "    'Mast cell': ['CPA3', 'KIT', 'TPSAB1', 'TPSB2'],\n",
    "    'Monocyte': [\n",
    "        'CD11b', 'CD14', 'CD163', 'CD68', 'FCN1', 'HLA-DR', 'IL1B', 'IL1R2'\n",
    "    ],\n",
    "    'Myeloid cell': ['CD68', 'LYZ'],\n",
    "    'Natural killer cell': [\n",
    "        'CD56', 'GNLY', 'KLRB1', 'KLRD1', 'KLRF1', 'NCAM1', 'NKG7'\n",
    "    ],\n",
    "    'Neutrophil': ['CD11b', 'CD15', 'CD16', 'CD66b', 'CSF3R', 'FCGR3B'],\n",
    "    'Pericyte': ['PDGFRB'],\n",
    "    'Stromal cell': ['ACTA2', 'PDGFRA'],\n",
    "    'T cell': [\n",
    "        'CD2', 'CD3', 'CD3D', 'CD3E', 'CD3G', 'CD4', 'CD8', 'CD8A',\n",
    "        'CD8B', 'IL7R', 'LAG3', 'NKG7', 'TRAC'\n",
    "    ],\n",
    "    'Cancer stem cell': ['CD44', 'CD133', 'EpCAM', 'ALDH1A1']\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04699ec3-af65-4c4d-8610-ca3103f5771f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_normalized.obs.sample()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6244c13-ce5d-4231-913e-0983783056bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import scipy.sparse\n",
    "\n",
    "cluster_key = 'nn_30 leiden_0.01'\n",
    "\n",
    "adata_combined = adata_normalized.copy()\n",
    "# Step 0: Ensure sliceable matrix\n",
    "if not isinstance(adata_combined.X, np.ndarray) and not scipy.sparse.isspmatrix_csr(adata_combined.X):\n",
    "    adata_combined.X = adata_combined.X.tocsr()\n",
    "\n",
    "# Step 1: Get all marker genes that exist in the data\n",
    "all_marker_genes = sorted(set(g for genes in marker_dict.values() for g in genes))\n",
    "present_markers = [g for g in all_marker_genes if g in adata_combined.var_names]\n",
    "\n",
    "# Step 2: Compute variance across all cells\n",
    "X = adata_combined[:, present_markers].X\n",
    "if not isinstance(X, np.ndarray):\n",
    "    X = X.toarray()  # convert sparse to dense if needed\n",
    "gene_variance = X.var(axis=0)\n",
    "\n",
    "# Step 3: Filter genes with variance above threshold\n",
    "var_threshold = 0.001  # You can tune this\n",
    "variable_genes = [g for g, v in zip(present_markers, gene_variance) if v > var_threshold]\n",
    "\n",
    "# Step 4: Filter marker dict\n",
    "filtered_marker_dict = {\n",
    "    celltype: [g for g in genes if g in variable_genes]\n",
    "    for celltype, genes in marker_dict.items()\n",
    "}\n",
    "filtered_marker_dict = {k: v for k, v in filtered_marker_dict.items() if len(v) > 0}\n",
    "\n",
    "# Step 5: Plot\n",
    "sc.pl.dotplot(\n",
    "    adata_combined,\n",
    "    var_names=filtered_marker_dict,\n",
    "    groupby=cluster_key,\n",
    "    standard_scale='var',  # Normalize each gene\n",
    "    cmap='Reds',\n",
    "    dendrogram=True,\n",
    "    show=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e25ca51f-c0ac-46e2-b3d9-c52da646ba3d",
   "metadata": {},
   "source": [
    "# DE clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d54ddf-9a45-4103-a17d-5ddb429b1c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_normalized.obs.sample()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb28003-ce87-43d2-bd22-91ff46173ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Make sure your AnnData object has cluster labels\n",
    "cluster_key = \"nn_30 leiden_0.01\"  # change to your actual key, e.g., \"nn_30 leiden_0.5\"\n",
    "if cluster_key not in adata_normalized.obs:\n",
    "    raise ValueError(f\"{cluster_key} not found in adata.obs\")\n",
    "\n",
    "# Perform differential expression\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata_normalized,\n",
    "    groupby=cluster_key,\n",
    "    method='t-test_overestim_var',  # or 'wilcoxon', 'logreg'\n",
    "    use_raw=False,\n",
    "    n_genes=100,\n",
    ")\n",
    "\n",
    "# Plot top DE genes\n",
    "sc.pl.rank_genes_groups(adata_normalized, n_genes=20, sharey=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9775ba9e-1541-43dd-82b9-bd09254d02f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e97cc86-d056-453a-b41b-e580e9ed3c0c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
